### PR TITLE
[KEYCLOAK-11185] Fixed build with JDK 11

### PR DESCRIPTION
**Pull request by douglaspalmer**
Tested build with the Docker image using OpenJDK 8,9,10,11,12,13,14.

### CODE REVIEW

File: saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLSignatureUtil.java [hunk 1]
```diff
@@ -41,7 +41,7 @@
 import org.xml.sax.SAXException;
 
 import javax.xml.bind.JAXBException;
-import javax.xml.crypto.MarshalException;
+import javax.xml.crypto.*;
```
Comment (abstractj) [inline saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLSignatureUtil.java]: @douglaspalmer I'd suggest to use explicit imports here, as it was before. Probably this is just some IDE configuration.

Comment (douglaspalmer) [reply saml-core/src/main/java/org/keycloak/saml/processing/core/util/XMLSignatureUtil.java]: ok, I'll fix that, (IntelliJ did it).

**Review by abstractj — DISMISSED** [review]

**Comment by abstractj** [comment]
@hmlnarik @stianst @mposolda because it touches the POM file and changes it a bit. If you guys have any thoughts about this change, please let us know. We are aiming to merge it before the end of the sprint.

**Review by mposolda — COMMENTED** [review]
@douglaspalmer @abstractj Will the build still work with IBM JDK and all the current java versions we support? That's my only concern/question regarding this.

**Comment by stianst** [comment]
LGTM, but resolve comments from @mposolda and @hmlnarik before merging

**Comment by vmuzikar** [comment]
I believe it would make sense to run both universal test pipelines with this. I know there are some instabilities / broken tests but at least we should check if it failed as usual.
I've manually triggered KC build pipeline with PEID "test-jdk11".

**Comment by vramik** [comment]
I've executed the build-pipeline as well :) peid: 65dfad25

**Comment by vmuzikar** [comment]
> I've executed the build-pipeline as well :) peid: 65dfad25

Ah, right – I should've check that first. ;) Aborted my run.

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@douglaspalmer I've tried to build the project with Oracle JDK 8, 11, 12 and 13. (Didn't try other vendors as I'm not sure which are even supported for building as @mposolda mentioned.)

Unfortunately, I've got the following compilation errors for JDK 11, 12 and 13.
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0-jboss-2:testCompile (default-testCompile) on project keycloak-saml-adapter-core: Compilation failure: Compilation failure: 
[ERROR] /Users/vmuzikar/Documents/keycloak/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java:[54,48] incompatible types: java.util.List<javax.xml.crypto.XMLStructure> cannot be converted to java.util.List<java.lang.Object>
[ERROR] /Users/vmuzikar/Documents/keycloak/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java:[55,48] incompatible types: java.util.List<javax.xml.crypto.XMLStructure> cannot be converted to java.util.List<java.lang.Object>
[ERROR] /Users/vmuzikar/Documents/keycloak/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java:[57,43] incompatible types: java.util.List<javax.xml.crypto.XMLStructure> cannot be converted to java.util.List<java.lang.Object>
[ERROR] /Users/vmuzikar/Documents/keycloak/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java:[60,44] incompatible types: java.util.List<javax.xml.crypto.XMLStructure> cannot be converted to java.util.List<java.lang.Object>
[ERROR] /Users/vmuzikar/Documents/keycloak/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java:[62,57] incompatible types: java.util.List<capture#1 of ?> cannot be converted to java.util.List<java.lang.Object>
[ERROR] /Users/vmuzikar/Documents/keycloak/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java:[68,48] incompatible types: java.util.List<javax.xml.crypto.XMLStructure> cannot be converted to java.util.List<java.lang.Object>
[ERROR] /Users/vmuzikar/Documents/keycloak/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java:[69,48] incompatible types: java.util.List<javax.xml.crypto.XMLStructure> cannot be converted to java.util.List<java.lang.Object>
[ERROR] /Users/vmuzikar/Documents/keycloak/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java:[71,43] incompatible types: java.util.List<javax.xml.crypto.XMLStructure> cannot be converted to java.util.List<java.lang.Object>
[ERROR] /Users/vmuzikar/Documents/keycloak/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java:[74,44] incompatible types: java.util.List<javax.xml.crypto.XMLStructure> cannot be converted to java.util.List<java.lang.Object>
[ERROR] /Users/vmuzikar/Documents/keycloak/adapters/saml/core/src/test/java/org/keycloak/adapters/cloned/HttpAdapterUtilsTest.java:[76,57] incompatible types: java.util.List<capture#2 of ?> cannot be converted to java.util.List<java.lang.Object>
```

So far I've build only the modules that are enabled by default. It's possible there're more compilation errors in other modules that are enabled by some profile (e.g. Admin Console UI tests). i need to figure out if we can somehow build all modules.

**Comment by vmuzikar** [comment]
@vramik I've just realized, running the test pipelines might not tell us much. KC build pipeline uses JDK 8 anyway, so we don't actually test if KC built with JDK >11 works fine. Am I right?

File: saml-core/src/main/java/org/keycloak/saml/processing/api/util/KeyInfoTools.java [hunk 2]
```diff
@@ -34,7 +35,7 @@
      * @param clazz
      * @return The object or {@code null} if not found.
      */
-    public static <T> T getContent(Iterable<Object> objects, Class<T> clazz) {
+    public static <T> T getContent(List objects, Class<T> clazz) {
```
Comment (hmlnarik) [inline saml-core/src/main/java/org/keycloak/saml/processing/api/util/KeyInfoTools.java]: Why is `Iterable` changed to `List`? Please revert

**Comment by douglaspalmer** [comment]
Iterable is changed to List to remove the generics error when compiling with jdk 9+.

**Comment by douglaspalmer** [comment]
@vmuzikar I think I've fixed the errors with HttpAdapterUtilsTest. Can you please take another look?

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@douglaspalmer Thanks for the update. Unfortunately, I can see another error with Oracle JDK >8:
```
[ERROR] Failed to execute goal on project keycloak-as7-adapter-spi: Could not resolve dependencies for project org.keycloak:keycloak-as7-adapter-spi:jar:8.0.0-SNAPSHOT: Could not find artifact com.sun:tools:jar:1.6 at specified path /Library/Java/JavaVirtualMachines/jdk-13.jdk/Contents/Home/../lib/tools.jar -> [Help 1]
```

**Comment by douglaspalmer** [comment]
> @douglaspalmer Thanks for the update. Unfortunately, I can see another error with Oracle JDK >8:
> 
> ```
> [ERROR] Failed to execute goal on project keycloak-as7-adapter-spi: Could not resolve dependencies for project org.keycloak:keycloak-as7-adapter-spi:jar:8.0.0-SNAPSHOT: Could not find artifact com.sun:tools:jar:1.6 at specified path /Library/Java/JavaVirtualMachines/jdk-13.jdk/Contents/Home/../lib/tools.jar -> [Help 1]
> ```

Are these errors caused by this PR or are they additional jdk problems we need to fix? If it's the later can we open a JIRA for the Oracle JDK issues?

**Comment by vmuzikar** [comment]
@douglaspalmer I'm not entirely sure what you mean by "additional jdk problems". If I understand it correctly, this PR was supposed to fix all the errors related to building with JDK higher than 8. And I believe this is one of the errors as it's not present with JDK 8.

On the other hand, the original JIRA mentions only OpenJDK and not Oracle JDK (which we AFAIK support as well). So maybe we could create a related JIRA for Oracle. But from my PoV there's no point in splitting this – we need to fix Oracle either way if we're to support newer JDKs in general.

Just one small correction. I've just now re-checked it and this error
```
[ERROR] Failed to execute goal on project keycloak-as7-adapter-spi: Could not resolve dependencies for project org.keycloak:keycloak-as7-adapter-spi:jar:8.0.0-SNAPSHOT: Could not find artifact com.sun:tools:jar:1.6 at specified path /Library/Java/JavaVirtualMachines/jdk-13.jdk/Contents/Home/../lib/tools.jar -> [Help 1]
```
Is present only with Oracle JDKs 12 and 13 – not 11 as I mistakenly stated before.

File: saml-core/src/main/java/org/keycloak/saml/processing/api/util/KeyInfoTools.java [hunk 3]
```diff
@@ -34,7 +34,7 @@
      * @param clazz
      * @return The object or {@code null} if not found.
      */
-    public static <T> T getContent(Iterable<Object> objects, Class<T> clazz) {
+    public static <T> T getContent(Iterable objects, Class<T> clazz) {
```
Comment (fperot74) [inline saml-core/src/main/java/org/keycloak/saml/processing/api/util/KeyInfoTools.java]: <Object> is not a good idea but as a best practice, you should not remove <Object>... May be List<?> would be better

Comment (hmlnarik) [reply saml-core/src/main/java/org/keycloak/saml/processing/api/util/KeyInfoTools.java]: `Iterable<?>` preferrably

Comment (fperot74) [reply saml-core/src/main/java/org/keycloak/saml/processing/api/util/KeyInfoTools.java]: Sorry (copied/pasted from first comment)

**Comment by abstractj** [comment]
> @douglaspalmer I'm not entirely sure what you mean by "additional jdk problems". If I understand it correctly, this PR was supposed to fix all the errors related to building with JDK higher than 8. And I believe this is one of the errors as it's not present with JDK 8.
> 
> On the other hand, the original JIRA mentions only OpenJDK and not Oracle JDK (which we AFAIK support for building as well). So maybe we could create a related JIRA for Oracle. But from my PoV there's no point in splitting this – we need to fix Oracle either way if we're to support newer JDKs in general.

@vmuzikar I had a chat with @stianst today and we reach an agreement that the issues with Oracle JDK should be fixed in the same PR. Could you please ask you to add all the details on that Jira? We can also change the title to JDK instead of OpenJDK.

**Comment by vmuzikar** [comment]
@douglaspalmer @abstractj I've updated the JIRA to reflect Oracle JDKs. There are really no additional details though. Just follow the steps to reproduce in the JIRA. [Here](https://gist.github.com/vmuzikar/7184433df37972c97c0671b4d4244745) you can find the error I mentioned with a bit more context.

**Comment by abstractj** [comment]
@vmuzikar thank you. I just tried to build this PR with Oracle JDK 11 and I got a successful build. Here's what I did to test it:
```
❯ java -version
java version "11.0.4" 2019-07-16 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.4+10-LTS)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.4+10-LTS, mixed mode)

❯ mvn --errors \
         --activate-profiles distribution \
         --projects distribution/server-dist,distribution/adapters/wildfly-adapter \
         --also-make -Dmaven.test.skip=true clean install
```
Also did:
```
❯ mvn clean install -Pdistribution -DskipTests=true
```

Is there anything I should do differently to see the errors you mentioned? For this PR we're not aiming JDK 12 or 13.

**Comment by vmuzikar** [comment]
@abstractj Please see the steps I took: https://gist.github.com/vmuzikar/d6cf41899f5a297ed9bfb0bf09d9ac2d

Just to be 100% sure it's not some weird macOS issue, I've also tried and reproduced it in a Fedora VM with the same result.

**Comment by abstractj** [comment]
> @abstractj Please see the steps I took: https://gist.github.com/vmuzikar/d6cf41899f5a297ed9bfb0bf09d9ac2d
> 
> Just to be 100% sure it's not some weird macOS issue, I've also tried and reproduced it in a Fedora VM with the same result.

I finally managed to reproduce your issue. It turns out that the Maven distribution installed through RPMs will come by default with JDK 8 set as runtime. It doesn't matter if you have `JAVA_HOME` set to JDK 11.

@douglaspalmer to see the issue you need to explicitly tell Maven to use JDK 11. Something like:

```
$ JAVA_HOME=/usr/lib/jvm/java-11-oraclejdk mvn clean install -Pdistribution -DskipTests=true
```

Even if you have JAVA_HOME set as env variable. At least here.

**Comment by vmuzikar** [comment]
@abstractj That's *really* weird. As I mentioned, I double-checked it using a Fedora VM. And there I've got a pretty standard maven installed from Fedora repos (i.e. RPM). All I had to do was to set `JAVA_HOME` env var to use a different JDK.

**Comment by hmlnarik** [comment]
This is apparently due to AS 7 not being compilable with Oracle JDK 11 due to https://github.com/jbossas/jboss-as/blob/7.2.0.Final/pom.xml#L282. Perhaps we should just remove the AS7 / EAP 6 modules when compiling with JDK > 8.

**Comment by vmuzikar** [comment]
@douglaspalmer I can see a new commit. Is this ready for re-review?

**Comment by douglaspalmer** [comment]
Yes please

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@douglaspalmer Thanks for the update. Unfortunately, I have some bad news – more compilation errors emerged.

**Oracle JDK 11 and higher**
"Keycloak Migration TestSuite For Server Config" is failing with the following compilation error.
```
[ERROR] Failed to execute goal org.wildfly.plugins:wildfly-maven-plugin:1.1.0.Final:start (start-stop-standalone) on project server-config-migration: Execution start-stop-standalone of goal org.wildfly.plugins:wildfly-maven-plugin:1.1.0.Final:start failed: Plugin org.wildfly.plugins:wildfly-maven-plugin:1.1.0.Final or one of its dependencies could not be resolved: Could not find artifact sun.jdk:jconsole:jar:jdk at specified path /Library/Java/JavaVirtualMachines/jdk-11.0.4.jdk/Contents/Home/../lib/jconsole.jar -> [Help 1]
```
This is however "just" a test module which is even disabled by default. Even though it's not ideal, we could fix this in a follow-up task.

**Oracle JDK 12 and higher**
"Keycloak Spring Security Integration" is failing with the following compilation error.
```
[INFO] Compiling 38 source files to /Users/vmuzikar/Documents/keycloak/adapters/oidc/spring-security/target/classes
[INFO] -------------------------------------------------------------
[WARNING] COMPILATION WARNING : 
[INFO] -------------------------------------------------------------
[WARNING] bootstrap class path not set in conjunction with -source 6
[INFO] 1 warning
[INFO] -------------------------------------------------------------
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] Source option 6 is no longer supported. Use 7 or later.
[ERROR] Target option 6 is no longer supported. Use 7 or later.
[INFO] 2 errors 
```
I haven't discovered this sooner because before that it never got to build this module as other errors appeared much sooner in other models (which skipped this module).

@douglaspalmer Please let me know if you need any help with verifying or reproducing this on your local machine. Perhaps we could schedule a brief call to resolve it more quickly.

**Comment by abstractj** [comment]
> **Oracle JDK 11 and higher**
> "Keycloak Migration TestSuite For Server Config" is failing with the following compilation error.
> 
> ```
> [ERROR] Failed to execute goal org.wildfly.plugins:wildfly-maven-plugin:1.1.0.Final:start (start-stop-standalone) on project server-config-migration: Execution start-stop-standalone of goal org.wildfly.plugins:wildfly-maven-plugin:1.1.0.Final:start failed: Plugin org.wildfly.plugins:wildfly-maven-plugin:1.1.0.Final or one of its dependencies could not be resolved: Could not find artifact sun.jdk:jconsole:jar:jdk at specified path /Library/Java/JavaVirtualMachines/jdk-11.0.4.jdk/Contents/Home/../lib/jconsole.jar -> [Help 1]
> ```
> 
> This is however "just" a test module which is even disabled by default. Even though it's not ideal, we could fix this in a follow-up task.

@vmuzikar thanks for reviewing this.

@hmlnarik @mposolda considering that this module is disabled by default. Do you guys think it is fine to follow up in another Jira, so we can move on with this change?

> 
> **Oracle JDK 12 and higher**
> "Keycloak Spring Security Integration" is failing with the following compilation error.

@vmuzikar despite the Jira description, is there any other reason to test this PR with JDK 12 or higher? The reason why I'm asking is because I would like us to focus only on JDK 11 get it done and later think about other versions if needed.

> 
> ```
> [INFO] Compiling 38 source files to /Users/vmuzikar/Documents/keycloak/adapters/oidc/spring-security/target/classes
> [INFO] -------------------------------------------------------------
> [WARNING] COMPILATION WARNING : 
> [INFO] -------------------------------------------------------------
> [WARNING] bootstrap class path not set in conjunction with -source 6
> [INFO] 1 warning
> [INFO] -------------------------------------------------------------
> [INFO] -------------------------------------------------------------
> [ERROR] COMPILATION ERROR : 
> [INFO] -------------------------------------------------------------
> [ERROR] Source option 6 is no longer supported. Use 7 or later.
> [ERROR] Target option 6 is no longer supported. Use 7 or later.
> [INFO] 2 errors 
> ```
> 
> I haven't discovered this sooner because before that it never got to build this module as other errors appeared much sooner in other models (which skipped this module).
> 
> @douglaspalmer Please let me know if you need any help with verifying or reproducing this on your local machine. Perhaps we could schedule a brief call to resolve it more quickly.

**Review by vmuzikar — APPROVED** [review]
@douglaspalmer Thank you for the update and for your patience with my constant complaints about this PR. :D Look good to me now!

**Review by abstractj — APPROVED** [review]

**Comment by mposolda** [comment]
Thanks,

I guess you already doublechecked that build still works with all the java versions, which we need to support, so that is not a concern anymore?

One more question: there are some dependencies introduced, so I wonder if we need to add some licences to the distribution https://github.com/keycloak/keycloak/tree/master/distribution/feature-packs/server-feature-pack/src/main/resources/licenses . I hope that no, but I don't have much clue about licences, so just asking if you are aware of this :)

**Review by hmlnarik — APPROVED** [review]
Technically looks good to me

**Review by abstractj — APPROVED** [review]

**Pull request merged**